File: app/src/main/java/com/example/todolist/TaskAdapter.java
package com.example.todolist;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import android.widget.TextView;

import com.example.todolist.Model.Task;

public class TaskAdapter extends RecyclerView.Adapter<TaskAdapter.ViewHolder> {

    interface OnTaskClickListener {
        void onTaskClick(Task task, int position);
    }

    private final OnTaskClickListener mOnClickListener;
    private final LayoutInflater mInflater;
    private final List<Task> mTasks;

    TaskAdapter(Context context, List<Task> tasks, OnTaskClickListener onClickListener) {
        this.mOnClickListener = onClickListener;
        this.mTasks = tasks;
        this.mInflater = LayoutInflater.from(context);
    }

//    public View getView(int position, View convertView, ViewGroup parent) {
//        final ViewHolder viewHolder;
//
//        if (convertView == null) {
//            convertView = inflater.inflate(this.layout, parent, false);
//            viewHolder = new ViewHolder(convertView);
//            convertView.setTag(viewHolder);
//        } else {
//            viewHolder = (ViewHolder) convertView.getTag();
//        }
//        return convertView;
//    }

    @NonNull
    @Override
    public TaskAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.task_layout, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(TaskAdapter.ViewHolder holder, @SuppressLint("RecyclerView") int position) {
        Task task = mTasks.get(position);
        holder.header.setText(task.getHeader());
        holder.description.setText(task.getDescription());
        holder.itemView.setOnClickListener(v -> mOnClickListener.onTaskClick(task, position));
    }

    @Override
    public int getItemCount() {
        return mTasks.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView header, description;

        //CheckBox checkbox;
        ViewHolder(View view) {
            super(view);
            header = view.findViewById(R.id.header);
            description = view.findViewById(R.id.description);
            //checkbox = view.findViewById(R.id.checkbox);

        }
    }
}

File: app/src/main/java/com/example/todolist/TaskAdapter.kt
package com.example.todolist

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.todolist.model.Task
import com.example.todolist.databinding.TaskLayoutBinding


class TaskAdapter(
    val onTaskClick: (task: Task) -> Unit
) : RecyclerView.Adapter<TaskAdapter.ViewHolder>() {

    private var tasks = emptyList<Task>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = TaskLayoutBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun getItemCount(): Int = tasks.size

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(tasks[position])
    }

    inner class ViewHolder(private val binding: TaskLayoutBinding) :
        RecyclerView.ViewHolder(binding.root) {
        private var item: Task? = null
        fun bind(taskItem: Task) {
            item = taskItem
            binding.header.text = taskItem.header
            binding.description.text = taskItem.description
        }

        init {
            itemView.setOnClickListener {
                item?.let { onTaskClick(it) }
            }
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setTasks(tasks: List<Task>){
        this.tasks = tasks
        notifyDataSetChanged()
    }
}